src/shogun/modelselection/NewGridSearch.cpp

```diff
@@ -0,0 +1,891 @@
+
+	for (auto param_pair = m_param_mapping.begin();
+	     param_pair != m_param_mapping.end(); ++param_pair)
+	{
+		ss << param_pair->first.c_str() << "=";
+		param_pair->second.visit(visitor.get());
+		if (std::next(param_pair) != m_param_mapping.end())
+			ss << ", ";
+	}
+
+	if (!m_nodes.empty())
+	{
+		if (!m_param_mapping.empty())
+			ss << ", ";
+
+		for (auto node = m_nodes.begin(); node != m_nodes.end(); ++node)
+		{
+			ss << (*node).first << "=";
+			if (node->second.size() > 1)
+			{
+				ss << "[";
+			}
+			for (auto inner_node = node->second.begin();
+			     inner_node != node->second.end(); ++inner_node)
+			{
+				(*inner_node)->to_string_helper(ss, visitor);
+				if (std::next(inner_node) != node->second.end())
+					ss << ", ";
+			}
+			if (node->second.size() > 1)
+			{
+				ss << "]";
+			}
+			if (std::next(node) != m_nodes.end())
+				ss << ", ";
+		}
+	}
+	ss << "))";
+	return ss.str();
+}
+
+void ParameterNode::to_string_helper(
+    std::stringstream& ss, std::unique_ptr<AnyVisitor>& visitor) const noexcept
+{
+	ss << get_parent_name() << "(";
+	for (auto param_pair = m_param_mapping.begin();
+	     param_pair != m_param_mapping.end(); ++param_pair)
+	{
+		ss << param_pair->first.c_str() << "=";
+		param_pair->second.visit(visitor.get());
+		if (std::next(param_pair) != m_param_mapping.end())
+			ss << ", ";
+	}
+	ss << ")";
+}
+
+bool ParameterNode::set_param_helper(const std::string& param, const Any& value)
+{
+	auto param_iter = param.find(delimiter);
+	SG_SDEBUG("ParameterNode::set_param_helper - param: %s\n", param.c_str())
+
+	if (param_iter != std::string::npos)
+	{
+		SG_SDEBUG(
+		    "ParameterNode::set_param_helper - Looking for param in nodes: "
+		    "%s\n",
+		    param.c_str())
+		for (auto& node : m_nodes)
+		{
+			if (node.second.size() > 1)
+			{
+				SG_SERROR(
+				    "Ambiguous call! More than one %s found in "
+				    "%s::%s. To specify a parameter in %s use "
+				    "the attach method\n",
+				    get_name(), get_parent_name(), param.c_str(), get_name())
+				return false;
+			}
+			else
+			{
+				SG_SDEBUG(
+				    "ParameterNode::set_param_helper - Looking for %s in %s\n",
+				    param.substr(param_iter + delimiter.size()).c_str(),
+				    node.first.c_str())
+				return node.second[0]->set_param_helper(
+				    param.substr(param_iter + delimiter.size()), value);
+			}
+		}
+	}
+	else
+	{
+		SG_SDEBUG(
+		    "ParameterNode::set_param_helper - Looking for %s in %s\n",
+		    param.c_str(), m_parent->get_name())
+		if (m_parent->has(param))
+		{
+			m_param_mapping.insert(std::make_pair(param, value));
+			return true;
+		}
+	}
+	return false;
+}
+
+ParameterNode* ParameterNode::get_current()
+{
+	auto tree = new ParameterNode(m_parent);
+
+	std::string param;
+
+	auto get_lambda = [&tree, &param](auto val) { tree->attach(param, val); };
+
+	for (const auto& param_pair : m_param_mapping)
+	{
+		param = param_pair.first;
+		sg_any_dispatch(
```

Comment:
cool that this is re-used here!

Reply:
yea, its really convenient, mostly because it defines a behaviour for 3 generic types: scalar, vector or matrix. whereas the whole visitor pattern would require writing all different scalar types, vector types and so on.. each time it is needed